### DISCUSSION TITLE
How to speed up or get fuzzy support for catching typos? 3.6 minutes vs. 3 seconds in tests

### DISCUSSION THREAD

**Discussion by kamranayub** [ravendb/ravendb#18684] [Q&A] [answered]
I have a fairly advanced document query with multiple subclauses to support searching video game titles + narrowing by platform or release date.

Here's an example query:

```
from index 'Games/ByMetadata' 
where boost(ExactSlug = 'fist-of-the-north-star', 150) or
      boost(fuzzy(ExactSlugs = 'fist-of-the-north-star', 0.9), 100) or
      
      boost(search(Slugs, 'fist north', and), 7) or
      boost(search(Slugs, 'fist north star', and), 50) or
      
      boost(search(Slugs, 'fist north star', and) and AvailablePlatforms in ('CachedPlatforms/1262-A'), 120) or
      boost(ExactSlug = 'fist-of-the-north-star' and AvailablePlatforms in ('CachedPlatforms/1262-A'), 500) or
      boost(fuzzy(ExactSlugs = 'fist-of-the-north-star', 0.9) and AvailablePlatforms in ('CachedPlatforms/1262-A'), 300) or
      boost(Release_ExpectedReleaseYear == 2010, 500) or
      
     search(Slugs, 'fist of the north star')
order by score()
```

As you can see I am using `fuzzy` twice. The `ExactSlugs` field is `FieldIndexing.Exact` and is an array of strings including the slugified game title + slugified alternative titles (a game may have 0..N alt names, usually less than 10). `Slugs` in the index is `FieldIndexing.Search`.

The query when executed individually is pretty "fast" at 234ms. Unfortunately, this is an order of magnitude slower when compared to _not_ using `fuzzy`, at <1ms.

And since my use case (bulk game import) calls for performing 1,100 searches _at once_ (using deferred queries), what ends up happening is that the whole operation takes about 3.6 minutes vs. 3 seconds in my tests. 

Is there any way of making `fuzzy` faster on my end? Or an alternative that will help me deal with minor typos in the same way?

I know Oren has mentioned `fuzzy` is very expensive. Is that always going to be the case?

Fuzzy helps catch very minor typos, which is very helpful for my use case because I want to provide the most accurate results on user-entered game titles, which can sometimes be off slightly. As a comparison, removing fuzzy search causes the algorithm to miss about 15 of my test cases. It's not the end of the world, but I would love to keep it in if I can. If my only option is to remove it, I guess it'll have to be. Fuzzy is not supported in Corax either (probably for this reason?).

**Comment by ayende** [comment, marked as answer]
Fuzzy is mostly about _user_ input, and the way it works is that we have to scan the _entire_ set of terms for the matches.

Let's say that you have 100K unique `ExactSlugs` - in order to compute `fuzzy`, we need to run over all of those terms and compute the Levenstien distance from your input. 

That is usually something that you'll do on a closed / small set of items. For example, names of countries or cities, not something as unique as slug.

Something that you may be able to do is to define `ExactSlugs` using ngram analyzer. That should give you much of the same behavior, far more cheaply.

**Reply by kamranayub** [reply]
Thanks, I was wondering about that. Is there a good example of using that?
